InputCommon: Dynamic Input Textures

iwubcode opened the pull request:
Per game input profiles make it easy to play a variety of Wii games.  However, due to the need to custom tailor input for most Wii games, it can be difficult to remember what input keys map to what.  It can be even more confusing when you haven't played a game in a while!  Some users have created texture packs that overwrite buttons in the UI with gamepad specific buttons but that can be frustrating because it forces a control scheme.  Enter "Dynamic Input Textures"!

Dynamic Input Textures allow texture creators to replace areas of a texture using a bunch of color boxes.  When the input keys are loaded, a custom texture gets generated with images representing the keys you have set!

Example [video](https://i.imgur.com/KW0lsPf.mp4)

Note: A big thank you to Jakey455 on the Dolphin forums for the XboxOne textures I used in my example

iwubcode commented:
There's also a "style" parameter.  Hoping to eventually introduce this feature to resource packs.  Users would be able to pick the style that fits their control scheme.  For instance, "XboxOne", "Xbox360", "WiiUPro", etc.

lioncash commented on Source/Core/InputCommon/DynamicInputTextureConfiguration.cpp:
```diff
@@ -0,0 +1,261 @@
+
+namespace
+{
+cv::Mat resizeKeepAspectRatio(const cv::Mat& input, const cv::Size& dstSize,
```
Our variable naming format is `snake_case`.

lioncash commented on Source/Core/InputCommon/DynamicInputTextureConfiguration.cpp:
```diff
@@ -0,0 +1,261 @@
+{
+  cv::Mat output;
+
+  double h1 = dstSize.width * (input.rows / (double)input.cols);
```
`h1` and `w2` can be const.  There's a few other occurrences within the function. also, `static_cast`.

lioncash commented on Source/Core/InputCommon/DynamicInputTextureConfiguration.cpp:
```diff
@@ -0,0 +1,261 @@
+{
+DynamicInputTextureConfiguration::DynamicInputTextureConfiguration(const std::string& json_file)
+{
+  std::ifstream json_stream(json_file);
```
You need to use `OpenFStream` within `Common/FileUtil.h`, otherwise this can fail to open on systems with different locales.

lioncash commented on Source/Core/InputCommon/DynamicInputTextureConfiguration.cpp:
```diff
@@ -0,0 +1,261 @@
+  std::ifstream json_stream(json_file);
+  if (!json_stream.is_open())
+  {
+    m_valid = false;
```
Was a return intended to follow this?

lioncash commented on Source/Core/InputCommon/DynamicInputTextureConfiguration.cpp:
```diff
@@ -0,0 +1,261 @@
+                       (std::istreambuf_iterator<char>()));
+
+  picojson::value out;
+  auto error = picojson::parse(out, json_str);
```
`const auto`

lioncash commented on Source/Core/InputCommon/DynamicInputTextureConfiguration.cpp:
```diff
@@ -0,0 +1,261 @@
+  }
+
+  // Required fields
+  picojson::value& colorkey_image = out.get("colorkey_image");
```
Many of these can be const references.

lioncash commented on Source/Core/InputCommon/DynamicInputTextureConfiguration.cpp:
```diff
@@ -0,0 +1,261 @@
+    m_style = style.to_str();
+  }
+
+  picojson::value& preserve_aspect_ratio = out.get("preserve_aspect_ratio");
```
const reference.

lioncash commented on Source/Core/InputCommon/DynamicInputTextureConfiguration.cpp:
```diff
@@ -0,0 +1,261 @@
+    return;
+  }
+
+  auto rgb_to_hsv = [](u8 r, u8 g, u8 b) {
```
```suggestion
  const auto rgb_to_hsv = [](u8 r, u8 g, u8 b) {
```

lioncash commented on Source/Core/InputCommon/DynamicInputTextureConfiguration.cpp:
```diff
@@ -0,0 +1,261 @@
+      continue;
+    }
+
+    auto input_image_iter = m_host_input_to_input_image.find(host_key);
```
```suggestion
    const auto input_image_iter = m_host_input_to_input_image.find(host_key);
```

lioncash commented on Source/Core/InputCommon/DynamicInputTextureConfiguration.cpp:
```diff
@@ -0,0 +1,261 @@
+    cv::Mat host_key_image =
+        cv::imread(m_base_path + input_image_iter->second, cv::IMREAD_UNCHANGED);
+
+    for (int i = 0; i < contours.size(); i++)
```
This can be a ranged for loop (which also avoids a signed/unsigned mismatch here).

lioncash commented on Source/Core/InputCommon/DynamicInputTextureConfiguration.h:
```diff
@@ -0,0 +1,39 @@
+class DynamicInputTextureConfiguration
+{
+public:
+  explicit DynamicInputTextureConfiguration(const std::string& json_file);
```
This class should likely default the destructor within the cpp file, given it contains several non-trivial types.

lioncash commented on Source/Core/InputCommon/DynamicInputTextureConfiguration.cpp:
```diff
@@ -0,0 +1,261 @@
+  }
+}
+
+void DynamicInputTextureConfiguration::GenerateTexture(IniFile::Section* sec)
```
`sec` can be a `const IniFile::Section*`, given it only queries the section contents. (ditto for `GenerateTextures`)

lioncash commented on Source/Core/InputCommon/DynamicInputTextureManager.h:
```diff
@@ -0,0 +1,21 @@
+#pragma once
+
+#include "Common/IniFile.h"
+#include "InputCommon/DynamicInputTextureConfiguration.h"
```
This can be replaced with a forward declaration if the constructor and destructor of `DynamicInputTextureManager` are defaulted into the cpp file.

iwubcode commented:
Thank you very much @lioncash .  I addressed those changes and  fixed two of the outstanding items (updated the configuration in the first post).  Still need to do some more testing and figure out how to get it to build on the buildserver!

iwubcode commented:
I did a [lot of testing](https://imgur.com/a/eiCkgK6) this weekend.  Games like Deadly Creatures, Go Vacation, etc use a bunch of smaller textures.  These are easier to "map" but require more json descriptor files, whereas Nintendo's own games often use a single texture for all of the controller icons.  One interesting thing to note for Spiderman Shattered Dimensions is all of the controller textures had a black background.  This seems like a colorkey but even making my replacement textures have a black background did not show the correct transparency.

Starmann commented:
This is one of the most creative pull requests I've seen in all my months of stalking Dolphin's pull requests for new features. I just have a few questions:

1. Have you tested it in Netplay? I'd imagine that could mess with things.

2. How would it handle the replacement of, say, the c-stick with arrow keys? What about a completely bizarre combination of keys that no sane person would choose?

3. Did you consider the possibility of this config being expanded for other dynamic forms of texture replacement when designing the format?

stenzek commented:
Netplay shouldn't be an issue, as all of the replacement is being done in VRAM. It'd only be an issue if an EFB copy was taken with a replaced texture, EFB2RAM was enabled, and the game was reading the copy with the CPU and made some decision based on the content which caused it to desync. Basically, it's the same as custom textures here.

Starmann commented:
@stenzek Yes but I'm wondering if the texture in question changes based on your controls in Netplay.

stenzek commented:
The replacement is done locally. So I imagine it would be per-player.

iwubcode commented:
@Starmann - let me try to answer some of your other questions.

>How would it handle the replacement of, say, the c-stick with arrow keys?

Assuming it's a single c-stick texture, you could add 4 colors to represent each input direction.  Then assuming the texture creator accounted for the arrow keys, it would just work.

>What about a completely bizarre combination of keys that no sane person would choose?

The texture creator has to define the button replacement textures.  If the button is not accounted for, that portion of the texture will be pulled from the "fallback texture".

One thing I decided not to deal with, is trying to build a texture dynamically based on the expression provided.  So button A & button B does not automatically create a texture with both keys in it.  However, if the texture creator thinks an expression is likely to be used (or is requested) they can map that to an image.

>Did you consider the possibility of this config being expanded for other dynamic forms of texture replacement when designing the format?

It's pretty hard wired to inputs at the moment.  But a similar idea could be implemented if there was a good use-case.  The thought did cross my mind.

Miksel12 commented:
Sounds like a fun idea but how are you going to manage mappings with multiple buttons? Eg 'Button 1' | 'Button 2'

iwubcode commented:
@Miksel12 

> how are you going to manage mappings with multiple buttons? Eg 'Button 1' | 'Button 2'

Like mentioned above, if the expression is supported in the pack, then it will just work.  However, in practice, I expect most texture creators will just map the normal buttons.  That means if you want to map shake to 'Button 1' | 'Button 2' you will be out of luck.  But if you want to map it to 'Button 3', it will work perfectly.

It's unfortunate but since expressions are evaluated at runtime, they don't fit well with the current implementation of Dynamic Input Textures which get generated when input configuration changes.

iwubcode commented:
Updated now to support multiple textures in a single configuration.  This can be useful if they all are going to use the same host controls.

I also found out why Spiderman Shattered Dimensions has odd black around it.  I believe there is another texture that causes the shadow to appear and that needs to be updated for the new icons.

I might test another game or two.  But the next step will be figuring out how to reduce openCV to the bare minimum needed to build.

iwubcode commented:
I moved away from opencv, so now all platforms build!  With that, I think it's ok to open to user testing and general code review.

phire commented:
Yeah, replacing a 100mb+ dependency with ~500 lines of code is very much the right move.

iwubcode commented:
Minor update, as I didn't fully understand the `RebuildUserDirectories` function when I originally looked at it.  Now if the "Load" directory gets rebuilt, we will properly set the dynamic texture directory.  I also rebased with latest master.

iwubcode commented:
Thanks to some testing from @jordan-woyak , some bugs were discovered with the implementation for Vulkan/OpenGL and also in general when a button gets 'cleared'.  The most recent push fixes those issues.

spycrab commented:
Please note that any format extension has to be documented in https://github.com/dolphin-emu/dolphin/blob/master/docs/ResourcePacks.md.

iwubcode commented:
I updated this PR taking much of the "documentation" from the first post and turning it into a markdown document.  I introduced "CustomFileFormats" directory which is expected to have markdown files for the various file formats that Dolphin supports.  I thought of calling it "Mods", if anyone has better ideas for a name let me know.

I updated my resource pack review to put the resource pack spec markdown there.  It also might be nice to put something about custom textures (since I think the only documentation is in the forums?) and game patches (#7982 ) if it gets merged.

BhaaLseN commented:
Whoops, I just commented on the  "CustomFileFormats" directory over in #8360 (since we already have the `ResourcePacks.md` in the `docs/` folder)

iwubcode commented:
I saw a lot of non-doc related files (rar, cpp, etc) which threw me off.  The documentation has been moved to `docs`

BhaaLseN commented on Source/Core/InputCommon/DynamicInputTextureConfiguration.cpp:
```diff
@@ -0,0 +1,296 @@
+        }
+        else
+        {
+          if (texture_data.m_preserve_aspect_ratio)
```
Is there any particular reason for this `else` not being joined with the `if` into an `else if`? There is no code directly inside this outer `else` (unless there should be, and it is missing?)

iwubcode replied:
Nope, just a style thing.  I fixed it!

iwubcode commented on Source/Core/InputCommon/ControllerEmu/ControllerEmu.h:
```diff
@@ -86,6 +88,8 @@ class EmulatedController
 
 private:
+  void GenerateTextures(IniFile::Section* sec);
+  InputCommon::DynamicInputTextureManager* m_dynamic_input_tex_config_manager = nullptr;
```
I've been thinking about this for a couple days and don't have any ideas for a better design.  I'm all ears if someone has an idea.  It should be safe though, since `InputConfig` owns both the `ControllerEmu` and the `DynamicInputTextureManager`.

iwubcode commented:
I was noticing at startup there was a flash of magenta.  It turns out this was because Dolphin was loading its texture, then loading input config (which told the textures to reload).  I've fixed it so that `HiResTexture` updates after input is loaded and I don't send a reload if inputs sees the CPU state is "Starting".

Additionally, I noticed if one configured Gamecube controllers and Wiimote with the same input, the texture generation might produce unexpected output.  The solution was to split up each controller configuration type into separate folders for the dynamic input texture configuration.  I updated the docs with this information.

Tilka commented:
What's the state of this PR?

iwubcode commented:
@Tilka - assuming no concerns, it should be ready to merge.  It had positive feedback from the users in the [forum thread](https://forums.dolphin-emu.org/Thread-introducing-dynamic-input-textures-a-new-feature-for-controller-icons).

jordan-woyak commented:
I'm trying to test this using the "Super Mario Galaxy" images/json you set me up with before but textures are not loading. Can you provide a zip or something to test with?

jordan-woyak commented:
Some things I ran into:
It's not immediately clear that "Load Custom Textures" needs to be enabled. The description of this setting still only mentions "User/Load/Textures"
Are separate "Wiimote"/"GCPad" directories the best organization method?
Would it make more sense to allow specifying all device types in one json file, especially since there could be a conflict of which texture to use when multiple input expressions mach?
Changing mappings causes a magenta flash but I guess master already does this when toggling "Load Custom Textures".

jordan-woyak commented on Source/Core/InputCommon/DynamicInputTextureConfiguration.cpp:
```diff
@@ -0,0 +1,302 @@
+    sec->Get(emulated_key, &host_key);
+
+    const auto color_only_mask = MaskColor(*color_key_image, Pixel{color.r, color.g, color.b, 255});
+    const auto rects = FindRectRegionsOfNeighboringPixels(color_only_mask);
```
You apply the copy operation to all found rectangles regardless of size.
I guess uses must ensure not one pixel of their chosen color is found in other parts of the image?

iwubcode replied:
Multiple replacements are a convenience at times.  I've seen a number of games that use the same sub-image (say an 'A' button icon) in multiple places in a single texture.

jordan-woyak commented on Source/Core/InputCommon/ImageOperations.h:
```diff
@@ -0,0 +1,77 @@
+{
+  u32 x = 0;
+  u32 y = 0;
+};
```
You could reuse an existing type we have.
`using Point = Common::TVec2<u32>;`

jordan-woyak commented on Source/Core/InputCommon/ImageOperations.h:
```diff
@@ -0,0 +1,77 @@
+  Point p;
+  u32 width = 0;
+  u32 height = 0;
+};
```
We also have a `Rectangle` class in MathUtil.
It uses top/left/right/bottom members though, so maybe you didn't want to work with that.

iwubcode replied:
Good call!  One disappointing thing is using `Rectangle<u32>` gives errors in VS because `std::abs` used in `GetHeight`/`GetWidth` is ambiguous.

jordan-woyak replied:
I should have fixed that recently. Do a rebase.

iwubcode replied:
In my local copy, I should be up-to-date with master.  Unfortunately, still the same error.

iwubcode replied:
I ended up keeping a similar signature to Rectangle with a note.

jordan-woyak commented on Source/Core/InputCommon/ImageOperations.h:
```diff
@@ -0,0 +1,77 @@
+  u32 height = 0;
+};
+
+void CopyTo(const ImagePixelData& src, ImagePixelData& dst, const Rect& src_region,
```
The name of this function could be more descriptive.
`CopyImageRegion` maybe?

jordan-woyak commented on Source/Core/VideoCommon/RenderBase.h:
```diff
@@ -403,6 +406,8 @@ class Renderer
 
   std::unique_ptr<NetPlayChatUI> m_netplay_chat_ui;
+
+  bool m_force_reload_textures = false;
```
It looks like this is written from by the host thread in `ForceReloadTextures` and read by the video thread in `Swap` without any sort of fencing.
It should probably be an `atomic<bool>`.

jordan-woyak commented on Source/Core/VideoCommon/TextureCacheBase.cpp:
```diff
@@ -137,6 +137,15 @@ void TextureCacheBase::Invalidate()
+
+  // Force hi res textures to reload
+  HiresTexture::Update(true);
+  HiresTexture::Update(false);
```
Why do you need to `Update` again after the "forced clear"?
If that is really needed this seems like an unintuitive interface.

iwubcode replied:
One update clears the hi resolution cache.  The second updates it with the new values.  I will add a new function to make things clearer (not sure why I did this).

jordan-woyak commented on docs/DynamicInputTextures.md:
```diff
@@ -0,0 +1,140 @@
+# Dolphin Dynamic Input Textures Specification (v1)
```
It should be mentioned somewhere to enable "Load Custom Textures".

iwubcode commented:
@jordan-woyak  - thank you for reviewing.  I updated with your feedback.

> It's not immediately clear that "Load Custom Textures" needs to be enabled. The description of this setting still only mentions "User/Load/Textures"

This was an oversight and I now mention both locations in the setting description.

> Are separate "Wiimote"/"GCPad" directories the best organization method?
Would it make more sense to allow specifying all device types in one json file, especially since there could be a conflict of which texture to use when multiple input expressions mach?

I can't recall why I had them separated.  However, thinking about your suggestion, it makes sense.  Even games that support multiple control types will have separate icons for Gamecube/Wii controllers.

I removed this intermediate directory from code and updated the documentation.

iwubcode commented:
@jordan-woyak - I took a stab at updating the configuration to support multiple emulation controllers and also multiple devices.  In all likelihood I expect most packs to be for a specific controller (PS4, Xbox360, etc) and maybe target a handful of devices or use the wildcard feature.

Here's an example of what the configuration looks like now:

```
{
	"preserve_aspect_ratio": true,
	"output_textures":
	{
		"tex1_128x128_02870c3b015d8b40_5.png":
		{
			"colorkey_image": "mario_galaxy_icons.png",
			"fallback_image": "mario_galaxy_icons_original.png",
			"emulated_controls":
			{
				"Wiimote1":
				{
					"Buttons/A": "#ff0000",
					"Buttons/B": "#ffff00"
				}
			},
			"host_controls":
			{
				"DInput/0/Keyboard Mouse":
				{
					"A": "xbox_one/a.png",
					"B": "xbox_one/b.png"
				}
			}
		}
	}
}
```

I also updated the documentation.  Describing a json map isn't easy!  Please let me know if you have any additional concerns.

jordan-woyak commented:
@iwubcode Yeah. I think the single json makes more sense.

iwubcode commented:
In order to make dynamic texture creation easier.  I wrote a tool in C#/WPF - https://github.com/iwubcode/DolphinDynamicInputTextureCreator .  I'm not an expert in tool design, so the community is free to create their own or fork off that.  Still a few features / bug-fixes coming before the first "official" release but it works right now.  See the tutorial at - https://github.com/iwubcode/DolphinDynamicInputTextureCreator/wiki/Tutorial for how to get started.

iwubcode commented:
Dynamic input textures now follow the same logic as #8628 , allowing them to be named whatever the creator likes.

iwubcode commented:
With #8628 merged, I now generate textures into a `<full gameid>_Generated` folder and place a `gameids` directory with the `<gameid>.txt` underneath it.  This should allow users who happen to have a region free directory with normal textures (that doesn't take advantage of #8628 ) still be loaded (in addition to any dynamically generated textures).

iwubcode commented:
@jordan-woyak - just curious if there were any other concerns preventing sign-off?

leoetlino commented on Source/Core/VideoCommon/RenderBase.h:
```diff
@@ -410,6 +413,8 @@ class Renderer
 
   std::unique_ptr<NetPlayChatUI> m_netplay_chat_ui;
+
+  std::atomic<bool> m_force_reload_textures = false;
```
It would be nice to use a `Common::Flag` here

leoetlino commented on Source/Core/InputCommon/DynamicInputTextureConfiguration.h:
```diff
@@ -0,0 +1,53 @@
+      u8 b;
+    };
+
+    std::string m_colorKeyImageName;
```
```suggestion
    std::string m_color_key_image_name;
```

ditto for the next two member names

leoetlino commented on Source/Core/InputCommon/DynamicInputTextureConfiguration.cpp:
```diff
@@ -0,0 +1,384 @@
+
+namespace
+{
+std::string get_stream_as_string(std::ifstream& stream)
```
```suggestion
std::string GetStreamAsString(std::ifstream& stream)
```

leoetlino commented on Source/Core/InputCommon/DynamicInputTextureConfiguration.cpp:
```diff
@@ -0,0 +1,384 @@
+  }
+
+  if (!any_dirty)
+  {
```
I would get rid of the braces here for readability.

leoetlino commented on Source/Core/InputCommon/DynamicInputTextureConfiguration.cpp:
```diff
@@ -0,0 +1,384 @@
+    const auto color_only_mask = MaskColor(*color_key_image, Pixel{color.r, color.g, color.b, 255});
+    const auto rects = FindRectRegionsOfNeighboringPixels(color_only_mask);
+
+    auto apply_fallback = [&]() {
```
```suggestion
    auto apply_fallback = [&] {
```

leoetlino commented on Source/Core/InputCommon/DynamicInputTextureConfiguration.cpp:
```diff
@@ -0,0 +1,384 @@
+      {
+        if (host_key_image->width == rect.GetWidth() && host_key_image->height == rect.GetHeight())
+        {
+          CopyImageRegion(*host_key_image, *color_key_image,
```
`Rect{0, 0, rect.GetWidth(), rect.GetHeight()}` and `dirty = true` can be moved out of the if cases.

iwubcode commented on Source/Core/InputCommon/DynamicInputTextureManager.h:
```diff
@@ -0,0 +1,26 @@
+
+#include <vector>
+
+#include "Common/IniFile.h"
```
This is not possible, since `Section` is a nested class

iwubcode commented:
@leoetlino - thank you very much for reviewing.  All comments have been addressed, please let me know if you see anything else

leoetlino commented on Source/Core/VideoCommon/HiresTextures.h:
```diff
@@ -14,11 +14,15 @@
 
 enum class TextureFormat;
 
+std::set<std::string> GetDirectoriesWithGameid(const std::string& root_directory,
```
```suggestion
std::set<std::string> GetDirectoriesWithGameId(const std::string& root_directory,
```

Actually, maybe keeping "texture" in the name (i.e. calling this GetTextureDirectoriesWithGameId) would be better? This is in the global namespace.

leoetlino commented on Source/Core/InputCommon/DynamicInputTextureConfiguration.cpp:
```diff
@@ -0,0 +1,375 @@
+      {
+        const std::string color_as_string = color.to_str();
+
+        bool color_code_error = false;
```
Could you extract the whole color parsing logic into its own function please? The function could return a `std::optional<int>` or even `std::optional<DynamicInputTextureData::ColorKey>`. I think that'd make this function easier to read.

iwubcode replied:
Good idea!

leoetlino commented on Source/Core/InputCommon/DynamicInputTextureConfiguration.cpp:
```diff
@@ -0,0 +1,375 @@
+    }
+    WriteImage(hi_res_folder + DIR_SEP + texture_data.m_hires_texture_name, *color_key_image);
+
+    const auto game_id_folder = hi_res_folder + DIR_SEP + "gamids";
```
`gameids`?

iwubcode replied:
Typo, thanks for catching!

leoetlino commented on Source/Core/InputCommon/ImageOperations.cpp:
```diff
@@ -0,0 +1,345 @@
+  if (info_ptr != nullptr)
+    png_free_data(png_ptr, info_ptr, PNG_FREE_ALL, -1);
+  if (png_ptr != nullptr)
+    png_destroy_write_struct(&png_ptr, (png_infopp) nullptr);
```
Is the cast required?

iwubcode commented:
@leoetlino - thank you again for your feedback.  All comments are addressed.  Please let me know if you see any other concerns.

leoetlino commented on Source/Core/VideoCommon/HiresTextures.h:
```diff
@@ -14,11 +14,15 @@
 
 enum class TextureFormat;
 
+std::set<std::string> GetTextureDirectoriesWithGameid(const std::string& root_directory,
```
```suggestion
std::set<std::string> GetTextureDirectoriesWithGameId(const std::string& root_directory,
```

(sorry for nitpicking)

iwubcode replied:
No problem at all.  Had to look for a while to see what was different :)

leoetlino commented on Source/Core/InputCommon/DynamicInputTextureConfiguration.cpp:
```diff
@@ -0,0 +1,371 @@
+    {
+      return std::stoi(color_code, 0, 16);
+    }
+    else if (color_code.size() == 8)
```
```suggestion
    if (color_code.size() == 8)
```

else isn't necessary after a return, and our convention is to get rid of those returns

iwubcode replied:
Thanks.  That's a good point but I'm going to keep it because if I use `TryParse` there's the chance that the parse will fail and if that happens I don't want to take the next branch.

leoetlino commented on Source/Core/InputCommon/DynamicInputTextureConfiguration.cpp:
```diff
@@ -0,0 +1,371 @@
+    const std::string color_code = str.substr(1);
+    if (color_code.size() == 6)
+    {
+      return std::stoi(color_code, 0, 16);
```
Apologies for missing this in my previous review, but I think we should avoid using std::stoi in Dolphin because it will throw exceptions for invalid values. TryParse (from our StringUtil) can be used as a replacement.

iwubcode replied:
Ah, yes, I wondered about that.  I'll make that change, thank you!

iwubcode replied:
Is there any built in Dolphin code to handle the requirement of converting to base 16?

leoetlino commented on Source/Core/InputCommon/DynamicInputTextureConfiguration.cpp:
```diff
@@ -0,0 +1,375 @@
+    {
+      // Color code includes alpha, that's currently not supported
+      int result;
+      if (TryParse(color_code.substr(2), &result, 16))
```
Wouldn't the alpha value be at the end of the string representation (RRGGBBAA)? If so, shouldn't this be `color_code.substr(0, 6)` instead?

iwubcode replied:
This was added based off [my tool](https://github.com/iwubcode/DolphinDynamicInputTextureCreator) which uses a microsoft color struct serialization.  Json generated via that tool had AA first from what I can tell.

iwubcode commented:
I hated to update the code after it was signed off on but I wanted to make this change before the configuration format got set in stone and users were using it out in the wild.

I originally had the pack creator specify a "colorkey" texture and a fallback texture.  The color key texture specified color regions and when combined with a mapping between emulated input-key and color, it provided all the necessary information to do the appropriate replacement.  My original thought was this was fairly simple to deal with.

However, users found the json cumbersome to write and the colorkey image, while easy was still a pain as you had to figure out what colors weren't in the original image!

I created [a tool](https://github.com/iwubcode/DolphinDynamicInputTextureCreator) to simplify writing the json but still the color mapping issue remained....

So, I removed the color specification in the json and replaced it with rectangle regions that mimic Dolphin's rect class.

This has a couple advantages:

* It's faster to process - Dolphin doesn't have to look for the regions by color, we can simply load the configuration up!
* It is more flexible - by using regions, we open up more possibilities.

The downside is, without the tool, it is cumbersome to define those regions.  I think it's a fair trade-off but I can revert if needed.

----

I also added the ability to specify the generated folder name.  It still defaults to `<gameid>_Generated` but gives some more flexibility.

----

@jordan-woyak  - please take another look at the code and let me know if you think this is a good design decision.  If so, I'll go ahead and squash the code so it's ready to be merged.

jordan-woyak reviewed (approved):
LGTM.

iwubcode commented:
Thank you @jordan-woyak for reviewing!  I squashed the commits, so this is now ready to go.  I'm putting the finishing touches on the tool.
